NetworkConfigurations/TrafficControlAndPolicingSubCategories/TrafficShaping.md

```diff
@@ -0,0 +1,506 @@
+- **rate**: The speed with which the queued traffic will be sent.
+- **burst**: The number of bytes of traffic that is dequeued before the shaper rate takes effect.
+
+### Hardware Statistics
+
+Hardware statistics are also supported. To see the statistics, use the show command that shows the current configuration, including TBF's statistics:
+
+```
+tc -s qdisc show dev <interface>
+```
+
+**Example Configuration:**
+
+The following example attaches a TBF qdisc under band 0 of an ETS parent whose handle is 101. It configures a 400Mbps shaper with a burst size of 128KiB:
+
+```
+tc qdisc add dev swp25 root handle 10: htb default 1
+tc class add dev swp25 parent 10: classid 10:1 htb rate 1Gbit
+
+tc qdisc add dev swp25 parent 10:1 handle 101: tbf rate 400Mbit burst 128K limit 1M
+```
+
+Output:
+
+```
+tc qdisc show dev swp25
+```
+
+```
+qdisc htb 10: root refcnt 2 r2q 10 default 0x1 direct_packets_stat 0 direct_qlen 1000
+```
+
+The qdisc htb output indicates the root queuing discipline's configuration, with values such as a reference count of 2, queuing limits set with a direct queue length of 1000 packets, and no packets sent or dropped, reflecting the current state of the root queue.
+
+```
+tc class show dev swp25
+```
+
+```
+class htb 10:1 root prio 0 rate 1Gbit ceil 1Gbit burst 1375b cburst 1375b
+```
+
+In the class htb output, the configured hierarchical token bucket class exhibits parameters like a priority of 0, a rate set to 1Gbit, and burst sizes of 1375 bytes for both the configured and committed burst, showcasing the root class's traffic shaping characteristics.
+
+```
+tc -s qdisc show dev swp25
+```
+
+```
+qdisc htb 10: root refcnt 2 r2q 10 default 0x1 direct_packets_stat 0 direct_qlen 1000
+ Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
+ backlog 0b 0p requeues 0
+qdisc tbf 101: parent 10:1 rate 400Mbit burst 131050b lat 18.4ms
+ Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0)
+ backlog 0b 0p requeues 0
+```
+
+The qdisc tbf output reveals specific details about the Token Bucket Filter settings under the root queuing discipline, including a rate of 400Mbit, a burst size of 131050 bytes, and a latency of 18.4 milliseconds, providing insights into traffic shaping at the root level.
+
+Verfication using `iperf` utility:
+
+```
+iperf -c 127.0.0.1 -t 60 -b 300M
+
+```
+
+Please make sure to run your iperf server using `iperf -s` command.
+
+Output-
+
+```
+------------------------------------------------------------
+Client connecting to 127.0.0.1, TCP port 5001
+TCP window size: 2.50 MByte (default)
+------------------------------------------------------------
+[  3] local 127.0.0.1 port 51824 connected with 127.0.0.1 port 5001
+[ ID] Interval       Transfer     Bandwidth
+[  3]  0.0-60.0 sec  2.10 GBytes   300 Mbits/sec
+```
+
+The test results indicate that the achieved bandwidth was indeed around 300 Mbits/sec, which is expected given that the traffic shaping is configured to limit the rate to 400 Mbit/s.
+
```

Comment:
Again, I do not believe this example proves you've limited the bandwidth. The limit is set to 400Mb/s, and you've sent 300Mb/s. Shouldn't you expect 300Mb/s?